openrtb_ext/request_wrapper.go

```diff
@@ -0,0 +1,808 @@
+	eids           *[]ExtUserEid
+	eidsDirty      bool
+}
+
+func (ue *UserExt) unmarshal(extJson json.RawMessage) error {
+	if len(ue.ext) != 0 || ue.Dirty() {
+		return nil
+	}
+	ue.ext = make(map[string]json.RawMessage)
+	ue.eids = &[]ExtUserEid{}
+	if len(extJson) == 0 {
+		return nil
+	}
+
+	if err := json.Unmarshal(extJson, &ue.ext); err != nil {
+		return err
+	}
+
+	consentJson, hasConsent := ue.ext["consent"]
+	if hasConsent {
+		if err := json.Unmarshal(consentJson, &ue.consent); err != nil {
+			return err
+		}
+	}
+
+	prebidJson, hasPrebid := ue.ext["prebid"]
+	if hasPrebid {
+		ue.prebid = &ExtUserPrebid{}
+		if err := json.Unmarshal(prebidJson, ue.prebid); err != nil {
+			return err
+		}
+	}
+
+	digiTrustJson, hasDigiTrust := ue.ext["digitrust"]
+	if hasDigiTrust {
+		ue.digiTrust = &ExtUserDigiTrust{}
+		if err := json.Unmarshal(digiTrustJson, ue.digiTrust); err != nil {
+			return err
+		}
+	}
+
+	eidsJson, hasEids := ue.ext["eids"]
+	if hasEids {
+		if err := json.Unmarshal(eidsJson, ue.eids); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ue *UserExt) marshal() (json.RawMessage, error) {
+	if ue.consentDirty {
+		consentJson, err := json.Marshal(ue.consent)
+		if err != nil {
+			return nil, err
+		}
+		if len(consentJson) > 0 {
+			ue.ext["consent"] = json.RawMessage(consentJson)
+		} else {
+			delete(ue.ext, "consent")
+		}
+		ue.consentDirty = false
+	}
+
+	if ue.prebidDirty {
+		prebidJson, err := json.Marshal(ue.prebid)
+		if err != nil {
+			return nil, err
+		}
+		if len(prebidJson) > 0 {
+			ue.ext["prebid"] = json.RawMessage(prebidJson)
+		} else {
+			delete(ue.ext, "prebid")
+		}
+		ue.prebidDirty = false
+	}
+
+	if ue.digiTrustDirty {
+		digiTrustJson, err := json.Marshal(ue.digiTrust)
+		if err != nil {
+			return nil, err
+		}
+		if len(digiTrustJson) > 0 {
+			ue.ext["digitrust"] = json.RawMessage(digiTrustJson)
+		} else {
+			delete(ue.ext, "digitrust")
+		}
+		ue.digiTrustDirty = false
+	}
+
+	if ue.eidsDirty {
+		if len(*ue.eids) > 0 {
+			eidsJson, err := json.Marshal(ue.eids)
+			if err != nil {
+				return nil, err
+			}
+			ue.ext["eids"] = json.RawMessage(eidsJson)
+		} else {
+			delete(ue.ext, "eids")
+		}
+		ue.eidsDirty = false
+	}
+	ue.extDirty = false
+
+	return json.Marshal(ue.ext)
+
+}
+
+func (ue *UserExt) Dirty() bool {
+	return ue.extDirty || ue.digiTrustDirty || ue.eidsDirty || ue.prebidDirty || ue.consentDirty
+}
+
+func (ue *UserExt) GetExt() map[string]json.RawMessage {
+	ext := ue.ext
+	return ext
+}
+
+func (ue *UserExt) SetExt(ext map[string]json.RawMessage) {
+	ue.ext = ext
+	ue.extDirty = true
+}
+
+func (ue *UserExt) GetConsent() *string {
+	if ue.consent == nil {
+		return nil
+	}
+	consent := *ue.consent
+	return &consent
+}
+
+func (ue *UserExt) SetConsent(consent *string) {
+	ue.consent = consent
+	ue.consentDirty = true
+}
+
+func (ue *UserExt) GetPrebid() *ExtUserPrebid {
+	if ue.prebid == nil {
+		return nil
+	}
+	prebid := *ue.prebid
+	return &prebid
```

Comment:
Hmm, a deep copy would be safer. It will eventually come down to is the work needed to make a deep copy happen worth the protection it gives us against bad coding? Any thoughts on this?